### ISSUE TITLE
Return `ListIterator` from `ArrayNode.elements()`

### ISSUE THREAD

**Issue by ludgerb** [FasterXML/jackson-databind#3120]
**Problem/Requirement:**
I have the requirement to traverse over Json document models and do selective manipulation on specific content by replacing some ValueNodes by other ValueNodes.

So the approach is, to traverse over the ObjectNode/ArrayNodes using the iterators provided via fields() (ObjectNode) respective elements() (ArrayNode).

Since for the ObjectNode the iterated object is the Map.Entry interface I can easily set a different JsonNode via entry.setValue(JsonNode value) without considering context:

```Java
Iterator<Map.Entry<String, JsonNode>> fields = currentNode.fields();
while (fields.hasNext()) {
  Map.Entry<String, JsonNode> entry = fields.next();
  entry.setValue( deriveNode(entry.getValue()) );
}
```

However, for ArrayNode, there is no approach to selectively replace single JsonNodes while iterating over the content.
To replace single fields, currently I need to:
1. Store the Parent ArrayNode
2. track the index I am working on
3. Call a parent.set(index, JsonNode)

Worse: the modification invalidates the iterator. Subsequent iteration calls will cause a ConcurrentModificationException.

Thus, after that, I need to re-obtain a new iterator instance and iterate index-n times to return to the previous position in order to continue from where I started.

**Proposed Solution**
A possible solution would be to internally use the ListIterator instead of an Iterator when calling ArrayNode.elements().
With a ListIterator, it is possible to cast the iterator to ListIterator and then call iterator.set(JsonNode value):

In ArrayNode.java:
```Java
@Override
public Iterator<JsonNode> elements() {
  return _children.listIterator();
 }
```

I do understand that changing the officially released method signature for the ArrayNode.elements() might cause compatibility issues therefore I do not request this.
Only under-the-hood-is the internally used instance could be of type listIterator, that would suffice. The user then has the potential to cast to ListIterator, since he has control over the specific version of the library used.

**Usage example**
With this small modification, the following can easily be done:
```Java
ListIterator<JsonNode> elements = (ListIterator<JsonNode>)currentNode.elements();
while (elements.hasNext()) {
  JsonNode childNode = elements.next();
  elements.set(deriveNode(childNode)); 
}
```

**Comment by cowtowncoder** [comment]
PR welcome -- against 2.18, and ideally ASAP if it's to go in 2.18.0.

Would need CLA from https://github.com/FasterXML/jackson/blob/master/contributor-agreement.pdf as well (usual way: print, fill & sign, scan/photo, email to `cla` at fasterxml dot com).

**Comment by JooHyukKim** [comment]
Went ahead and wrote the solution #4692.
Hope u like it, @ludgerb ! 👍🏼

**Comment by cowtowncoder** [comment]
Merged in for inclusion in 2.18.0
